### PR TITLE
Cleanups and refactorings

### PR DESCRIPTION
**Pull request by milseman**
- Split off matching code/logic/types from `DSL/Core.swift`
- Make a `Util/Errors.swift` for our `Unsupported` and `Unreachable` errors, and use them more broadly
- Split off AST conversion logic from `DSLTree.swift`

### CODE REVIEW

**Comment by milseman** [comment]
@rxwei I don't want to cause conflicts for your PRs, so this can wait until later to merge.

**Comment by milseman** [comment]
@swift-ci please test linux platform

File: Sources/_StringProcessing/RegexDSL/Core.swift
Comment (rxwei) [inline Sources/_StringProcessing/RegexDSL/Core.swift:255]: FWIW, #127 removes `MatchProtocol`, so you can probably leave it here so that there's one fewer conflict.

**Review by rxwei — APPROVED** [review]

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Comment by rxwei** [comment]
Feel free to merge this as there's no ETA when a new toolchain will come out to unblock my PRs.

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Pull request merged**
